### PR TITLE
Polygon/line creation interface completed

### PR DESCRIPTION
**Pull request by BarAgam**
Implemented send geometry (polylines and polygons) view-model logic.
You can now create and send polylines and polygons from the tablet.

Refactoring -
Introduced MapDrawer and MapEntityFactory to allow easy creation and manipulation of entities over the map.
These are used in the new code and in MeasureViewModel and SendQuadrilateralViewModel.

### CODE REVIEW

File: app/src/main/java/com/teamagam/gimelgimel/app/map/viewModel/SendGeometryViewModel.java [hunk 1]
```diff
@@ -2,20 +2,38 @@
...
 import com.teamagam.gimelgimel.domain.rasters.DisplayIntermediateRastersInteractorFactory;
+import java.util.ArrayList;
+import java.util.List;
+
+import static android.text.TextUtils.isEmpty;
 
 @AutoFactory
 public class SendGeometryViewModel extends BaseMapViewModel {
 
+  private static final String EMPTY_STRING = "";
```
Comment (teamagam5) [inline app/src/main/java/com/teamagam/gimelgimel/app/map/viewModel/SendGeometryViewModel.java:25]: Maybe we should have only one occurrence of an empty string in CONSTANTS...?

File: app/src/main/java/com/teamagam/gimelgimel/app/map/viewModel/MapEntityFactory.java [hunk 2]
```diff
@@ -0,0 +1,72 @@
...
+  }
+
+  public static class SimpleSymbolizer implements Symbolizer {
+    @Override
+    public PointSymbol createFromPoint(PointGeometry point) {
+      return new PointSymbol(false, PointSymbol.POINT_TYPE_CIRCLE);
+    }
+
+    @Override
+    public PolylineSymbol createFromPolyline(List<PointGeometry> points) {
```
Comment (teamagam5) [inline app/src/main/java/com/teamagam/gimelgimel/app/map/viewModel/MapEntityFactory.java]: It looks like the signature should be createFromPolyline(Polyline polyline) somehow

File: app/src/main/java/com/teamagam/gimelgimel/app/map/viewModel/MapEntityFactory.java [hunk 3]
```diff
@@ -0,0 +1,72 @@
...
+      return new PolylineSymbol(false);
+    }
+
+    @Override
+    public PolygonSymbol createFromPolygon(List<PointGeometry> points) {
```
Comment (teamagam5) [inline app/src/main/java/com/teamagam/gimelgimel/app/map/viewModel/MapEntityFactory.java]: Once again

File: app/src/main/java/com/teamagam/gimelgimel/app/map/viewModel/MapEntityFactory.java [hunk 4]
```diff
@@ -0,0 +1,72 @@
...
+  public MapEntityFactory() {
+    this(new SimpleSymbolizer());
+  }
+
+  public PointEntity createPoint(PointGeometry point) {
+    return new PointEntity(generateId(), EMPTY_STRING, point, mSymbolizer.createFromPoint(point));
+  }
+
+  public PolylineEntity createPolyline(List<PointGeometry> points) {
+    Polyline polyline = new Polyline(points);
```
Comment (teamagam5) [inline app/src/main/java/com/teamagam/gimelgimel/app/map/viewModel/MapEntityFactory.java]: Inline? (in the following method it is)

File: app/src/main/java/com/teamagam/gimelgimel/app/map/viewModel/SendGeometryViewModel.java [hunk 5]
```diff
@@ -24,21 +42,43 @@ protected SendGeometryViewModel(
           DisplayVectorLayersInteractorFactory displayVectorLayersInteractorFactory,
       @Provided
           DisplayIntermediateRastersInteractorFactory displayIntermediateRastersInteractorFactory,
-      GGMapView ggMapView) {
+      @Provided
+          SendGeoMessageInteractorFactory sendGeoMessageInteractorFactory, GGMapView ggMapView,
```
Comment (teamagam5) [inline app/src/main/java/com/teamagam/gimelgimel/app/map/viewModel/SendGeometryViewModel.java]: break line

File: app/src/main/java/com/teamagam/gimelgimel/app/map/viewModel/SendGeometryViewModel.java [hunk 6]
```diff
@@ -24,21 +42,43 @@ protected SendGeometryViewModel(
...
+      InvalidInputNotifier invalidInputNotifier, ViewDismisser viewDismisser) {
```
Comment (teamagam5) [inline app/src/main/java/com/teamagam/gimelgimel/app/map/viewModel/SendGeometryViewModel.java]: br

File: app/src/main/java/com/teamagam/gimelgimel/app/map/viewModel/SendGeometryViewModel.java [hunk 7]
```diff
@@ -53,5 +93,82 @@ public String getDescription() {
...
+
+  private boolean isPolylineState() {
+    return !mIsSwitchChecked;
+  }
+
+  private void displayPolygon(List<PointGeometry> points) {
+    mMapDrawer.draw(mMapEntityFactory.createPolygon(points));
+  }
+
+  public interface InvalidInputNotifier {
```
Comment (teamagam5) [inline app/src/main/java/com/teamagam/gimelgimel/app/map/viewModel/SendGeometryViewModel.java:170]: Isn't it a common behavior that SendQuad... and SendGeo... should share? I mean, the validation process- I think it belongs on both.

File: app/src/main/res/values/strings.xml [hunk 8]
```diff
@@ -100,6 +104,7 @@
 
   <!--Send Geometry -->
   <string name="send_geometry_toolbar_title">Send Geometry</string>
+  <string name="send_geometry_invalid_input_message" translatable="false">@string/send_quadrilateral_invalid_input_message</string>
```
Comment (teamagam5) [inline app/src/main/res/values/strings.xml]: I already said I think it's common to both Send...Fragments. Anyway, if not- this should point to @string/invalid_input_message no?

**Review by teamagam5 — APPROVED** [review]

File: app/src/main/java/com/teamagam/gimelgimel/app/map/viewModel/SendGeometryViewModel.java (hunk 5 above)
Comment (BarAgam) [reply app/src/main/java/com/teamagam/gimelgimel/app/map/viewModel/SendGeometryViewModel.java]: K

File: app/src/main/java/com/teamagam/gimelgimel/app/map/viewModel/MapEntityFactory.java (hunk 4 above)
Comment (BarAgam) [reply app/src/main/java/com/teamagam/gimelgimel/app/map/viewModel/MapEntityFactory.java]: K

File: app/src/main/java/com/teamagam/gimelgimel/app/map/viewModel/MapEntityFactory.java (hunk 3 above)
Comment (BarAgam) [reply app/src/main/java/com/teamagam/gimelgimel/app/map/viewModel/MapEntityFactory.java]: Resolved

**Pull request merged**